### ISSUE TITLE
FDSN Client  wadl, catalogs, and contributors query frequency

### ISSUE THREAD

**Issue by nick-iris** [obspy/obspy#2094]
When a FDSN Client is instantiated for the first time it makes a request for each web service WADL as well as the event services catalogs and contributors data. This is placing a heavy load on the database since collecting the information for the contributors and catalogs is not an efficient query.

Would it be possible to not request the event catalog and contributors information when a FDSN Client is first created, unless explicitly requested by the user?

**Example:**
Creating a client for the first time makes several wadl requests and a request for the event catalogs and contributors.
```
>>> c = Client("IRIS", debug=True)
Base URL: http://service.iris.edu
Request Headers: {u'User-Agent': u'ObsPy/1.0.3.post0+1835.g80139da8b6.obspy.fdsnws.routing (Darwin-16.7.0-x86_64-i386-64bit, Python 2.7.10)'}
Downloading http://service.iris.edu/fdsnws/dataselect/1/application.wadl with requesting gzip compression
Downloading http://service.iris.edu/fdsnws/event/1/application.wadl with requesting gzip compression
Downloading http://service.iris.edu/fdsnws/station/1/application.wadl with requesting gzip compression
Downloading http://service.iris.edu/fdsnws/event/1/catalogs with requesting gzip compression
Downloading http://service.iris.edu/fdsnws/event/1/contributors with requesting gzip compression
Uncompressing gzipped response for http://service.iris.edu/fdsnws/station/1/application.wadl
Uncompressing gzipped response for http://service.iris.edu/fdsnws/event/1/application.wadl
Uncompressing gzipped response for http://service.iris.edu/fdsnws/dataselect/1/application.wadl
Downloaded http://service.iris.edu/fdsnws/event/1/application.wadl with HTTP code: 200
Downloaded http://service.iris.edu/fdsnws/station/1/application.wadl with HTTP code: 200
Downloaded http://service.iris.edu/fdsnws/dataselect/1/application.wadl with HTTP code: 200
Downloaded http://service.iris.edu/fdsnws/event/1/catalogs with HTTP code: 200
Uncompressing gzipped response for http://service.iris.edu/fdsnws/event/1/contributors
Downloaded http://service.iris.edu/fdsnws/event/1/contributors with HTTP code: 200
Discovered station service
Discovered event service
Discovered dataselect service
Storing discovered services in cache.
```

These get cached so the next time a client is created it will use what was already loaded, but the initial request for the catalogs and contributors puts a heavy load on the database even though they often go unused.

```
>>> b = Client("IRIS", debug=True)
Base URL: http://service.iris.edu
Request Headers: {u'User-Agent': u'ObsPy/1.0.3.post0+1835.g80139da8b6.obspy.fdsnws.routing (Darwin-16.7.0-x86_64-i386-64bit, Python 2.7.10)'}
Loading discovered services from cache.
```

**Comment by krischer** [comment]
Yes this might indeed be a good idea. These are really not used a lot so some kind of lazy mechanism might be a good idea. I guess the cleanest backwards compatible way would be to write some custom dictionary for the `self.services` attribute that would lazily download the catalogs and contributors when required.

Nonetheless I think this should also be fixed server side. These are quasi-static XML files so some kind of cache in front of the server (e.g. varnish) could just serve these without having the touch the database at all making them almost free to serve like the WADL files.

**Comment by megies** [comment]
I guess changing this to a lazy lookup could be seen as a non-feature change, so I'll put this on 1.1.x. Feel free to change, though.

**Comment by jkmacc-LANL** [comment]
I think I've just run into this, too.  I'm initializing tons of `Client` objects, and each makes one of these services queries to IRIS, resulting in a `ValueError: The current client does not have a dataselect service.`, presumably because the database can't keep up.  I'm not sure, but I suspect this is because of what @nick-iris mentioned.  I don't know if I'd prefer a `cache_services=False` or a `query_services=True` init keyword.  The latter is useful when you *know* what your endpoint has, and you're not interested in the overhead.  Maybe both?

**Comment by d-chambers** [comment]
Would lazifying the wadl requests also allow for instantiating a `client` object without an internet connection?

**Comment by jkmacc-LANL** [comment]
Hmm....  [This seems to be the bit that gets executed.](https://docs.obspy.org/_modules/obspy/clients/fdsn/client.html#Client._discover_services)  You can't initialize one without the internet?  I guess that makes sense...

**Comment by jkmacc-LANL** [comment]
@d-chambers Why would you want to init a `Client` without the internet?  Testing?

**Comment by d-chambers** [comment]
We (of course) use obspy extensively in our data pipeline. There are functions that , upon receiving event data, attempt to pull waveforms for processing. The functions that do this first check to see if the event data is stored locally, and if not reach out to IRIS to get the data found there. Consequently, we initialize a single IRIS client that all other functions can use. It is most natural to simply initialize the client at import time but often the client instance isn't used and it becomes problematic when working on unreliable internet connections, as we often do. Of course there are ways around this, but it would be cleaner if the Clients were lazy. 

Lazy connections to external data sources are generally good practice. Most database drivers I have used behave this way by default.

**Comment by jkmacc-LANL** [comment]
I'm interested in making `client.services` optional, not just lazy.  I know IRIS DMC has `dataselect` and I'm not interested in asking them every time I init a `Client`.  I think `client.services` a useful thing for learning about or querying multiple endpoints, but I think it should be possible to disable it.  

Is there a PR for this?  If not, may I start one that adds some sort of `Client(..., query_services=True)` keyword?

**Comment by krischer** [comment]
The autodetection of an FDSN node's capabilities is fairly deeply ingrained into everything. Would making it a lazy object work for you guys? We could make it lazy all the way down, i.e. a dataselect WADL is only called when the first dataselect call is issues and so on.

All the service calls to the WADLs and what not should already be cached per Python processes. I do agree that it is an issue when one of the initial calls fails for some transient issue.

**Comment by krischer** [comment]
I'm not sure what would be gained by fully disabling it.

**Comment by nick-iris** [comment]
I believe making them lazy would work for our needs, but what happens if an FDSN web service endpoint is missing a WADL. Will the client completely fail? If it does fail then I can see some benefit of making `client.services` optional as @jkmacc-LANL suggests.

**Comment by krischer** [comment]
At that point it is no longer a valid FDSN web service and I'm not sure we want to cater to that case. Even if just for development purposes it is quite simple to just add the WADL route to the server and serve some default file.

It would always be possible to subclass the `Client` class and replace the `.services` with some mock object if so desired. If that is indeed a more common use case I'd be fine with offering this derived class but I'm not convinced we want this added complexity as part of ObsPy.

**Comment by jkmacc-LANL** [comment]
Would making the WADL query lazy help the problem @nick-iris has?  In looking at the code, `.services` is used by `.get_stations`, `.get_events`, and `.get_waveforms`, and it is *already* cached.  It seems that pushing the WADL query down to the method calls won't save any WADL queries, since those methods will most certainly be called.

**Comment by jkmacc-LANL** [comment]
@krischer Fully disabling the WADL query would be useful when you already know what services and parameters your endpoint provides, and you just want to do a ton of requests on them without swamping them with WADL stuff.

**Comment by krischer** [comment]
> Would making the WADL query lazy help the problem @nick-iris has? In looking at the code, `.services` is used by `.get_stations`, `.get_events`, and `.get_waveforms`, and it is _already_ cached. It seems that pushing the WADL query down to the method calls won't save any WADL queries, since those methods will most certainly be called.

Right now the WADLs are hit at init time. If it is lazy they would be hit at calling the actual `.get_stations()` and so on webservices. I think this would completely alleviate the issue.


> @krischer Fully disabling the WADL query would be useful when you already know what services and parameters your endpoint provides, and you just want to do a ton of requests on them without swamping them with WADL stuff.

I don't think this would help. The main reason is that all the actual calls are dynamic and require work from the actual fdsnws implementation. The WADLs are hopefully cached higher up the stack by something like varnish and thus serving them is basically free. The only price to pay would be a slight increase in latency before the very first data request can actually hit the server. I think that is an okay trade-off. And we gain all the validation we can do before even hitting the remote server. We don't only use the WADLs to discover what services are there but also what each endpoint offers.

**Comment by chad-iris** [comment]
> The WADLs are hopefully cached higher up the stack by something like varnish and thus serving them is basically free.

This is not free and assumes a lot about the flexibility and capacity of service providers.  If a data collection exercise is known to require a million requests and for some reason each one needs an instantiation, it would be _really_ nice to be able to turn off the needless WADL requests.  No matter how cheap implementors make them it's more than zero and a waste for all involved.

I think it would be good for the Client to be able to turn them off.  In such a case, or when the WADL retrieval fails, a minimal level of functionality defined by the required arguments is assumed.

**Comment by krischer** [comment]
My hope would be that this worst case scenario does not happen at least with ObsPy as the WADLs are cached at the module level. Only the very first instantiation of a client object to a certain data center will actually download something - every other instantiation would just reuse them and the cache would remain active as long as the Python process is alive. But it could happen if a new Python process is started for every request - I don't know if someone would write the required loops outside of Python.

Assuming we make the WADL calls lazy, the WADL requests would only actually be triggered immediately before the very first request to one of the webservice (and only once per Python process).  It's hard for me to imaging a scenario where a WADL request is not *a lot* cheaper than an actual (meta)data request for the server. At the very least (and even if the WADL files are read from disc every request) they would still be in the OS provided disc caches.

The ObsPy client gains a fair bit of flexibility by using the WADL files and can thus can for example easily deal with client side validation and discoverability of non-default parameters. I think we want to keep this in some fashion.

But if everyone agrees that this an issue maybe we can work out something that works for everyone? Offering the option to optionally not use the WADLs would IMHO not help. Everyone who realizes that this is an issue has the knowledge to just derive a child class that deals with it.

One possibility I can imaging right now is that we only parse the WADL files if:

* A user calls a service with a non-standard parameter or we require the WADL for some other reason.
* If the web service responds with a status code which we can't immediately make sense of.

Otherwise just assume the service implements the vanilla specs.

**Comment by chad-iris** [comment]
The possibility suggested by @krischer seems reasonable.

While I think it's a better general strategy to be more lenient, meaning less reliant on a WADL, I do not think there is a broad problem, at least at the DMC.  Scraping through the last few days of DMC service logs, we get 40,000+ WADL requests/day from ObsPy (across all services), so once every 2 seconds on average.  Each day there are between 100 and 150 unique ObsPy clients accessing data.  Not a problem at those counts, but that's a lot of WADL access that probably made no difference beyond reconfirming that, yep, it looks like an FDSN service.

**Comment by chad-iris** [comment]
Also not to forget the `catalogs` and `contributors` queries to the fdsnws-event service that are performed even if the event service is unused as @nick-iris originally wrote.

Each instantiation of `Client("IRIS")` makes 5 connections to IRIS (3 WADLs and the event specific calls).  My guess is that most of the information is unnecessary most of the time.

**Comment by megies** [comment]
I guess we might try to sort this one out for 1.2.0

**Comment by jkmacc-LANL** [comment]
> Right now the WADLs are hit at init time. If it is lazy they would be hit at calling the actual .get_stations() and so on webservices. I think this would completely alleviate the issue.

Is the idea that only the WADL relevant to `.get_waveforms`, for example, would be queried by that method when it is used, instead of all WADLs for all methods at init time?  I guess this would reduce requests for unused services.  Gotcha.

> Only the very first instantiation of a client object to a certain data center will actually download something - every other instantiation would just reuse them and the cache would remain active as long as the Python process is alive. But it could happen if a new Python process is started for every request - I don't know if someone would write the required loops outside of Python.

While this certainly isn't common, I currently have a distributed computation that instantiates `Client` objects at a rate of 60-120 *per second* (on separate python processes I think), and this can go on for hours.  If each of these pings the DMC 5 times, that's a lot of unnecessary work for the DMC and it slows down the application.  I do get `ValueError: The current client does not have a dataselect service.` exceptions, and I suspect it's because of WADL queries.  It seems reasonable to assume certain things about endpoints from a familiar provider like the DMC, and to have a `discover_services=False` option to streamline the request.  Something like:

```python
class Client(object):
    ...
    def __init__(self, ..., discover_services=True):
        ...
        self.discover_services = discover_services
        ...
        if self.discover_services:
            self._discover_services()
        else:
            self.services = {}
        ...

    def get_waveforms(self, ...):
        ...
        if self.discover_services and "event" not in self.services:
                msg = "The current client does not have an event service."
                raise ValueError(msg)
        ....
```

**Comment by d-chambers** [comment]
> I currently have a distributed computation that instantiates `Client` objects at a rate of 60-120 _per second_ (on separate python processes I think), and this can go on for hours.

Wow! Sounds like IRIS might solve a lot of problems by just banning @jkmacc-LANL  :wink:

**Comment by chad-iris** [comment]
> Wow! Sounds like IRIS might solve a lot of problems by just banning @jkmacc-LANL 😉

Yeah, he'd just find another set of IPs to use, becoming a game of whack-a-mole 😜

But the DMC does block folks that make too many connections too fast using a load balancer as a gateway.  On the web service backend we don't see that traffic.  What the client sees is failed connections during a short ban period.

**Comment by jkmacc-LANL** [comment]
To be clear, this was a limited-time test and done with DMC knowledge 😉   But also, I think it exposed that the distributed use case for the `Client` was not originally anticipated.  Could I offer a PR (for 1.2.0?) to implement the shortcut outlined above?  (I'm fuzzier on how to do the lazy query part that's been discussed.)

**Comment by krischer** [comment]
With #2325 merged there is now a temporary workaround to this solution available to power users:

```python
[ins] In [2]: c = Client("IRIS", debug=True, _discover_services=False)
Installed new opener with handlers: [<obspy.clients.fdsn.client.CustomRedirectHandler object at 0x108c957f0>]
Base URL: http://service.iris.edu
Request Headers: {'User-Agent': 'ObsPy/1.1.1.post0+584.g4d35268832 (Darwin-18.2.0-x86_64-i386-64bit, Python 3.7.1)'}
```

I'm puning this to the next release because we long term do want a proper lazy solution that just does the right thing and would be used by everyone.

**Comment by crotwell** [comment]
Just my $0.02, but I would consider making `_discover_services=False` the default in the next release. It can save quite a bit of time at startup for the client, especially given 99.9% of queries are probably just using one of the standard data centers. 

Ask for forgiveness not permission! :)
